Common/src/interface_interpolation/CInterpolator.cpp

```diff
@@ -171,110 +171,117 @@ void CInterpolator::ReconstructBoundary(unsigned long val_zone, int val_marker){
 
-  unsigned long iVertex, jVertex, kVertex;
+  unsigned long iVertex, kVertex;
 
-  unsigned long count, iTmp, *uptr, dPoint, EdgeIndex, jEdge, nEdges, nNodes, nVertex, iDim, nDim, iPoint;
+  unsigned long *uptr, nVertex, nElems, iDim, nDim, iPoint;
 
   unsigned long nGlobalLinkedNodes, nLocalVertex, nLocalLinkedNodes;
 
   nDim = geom->GetnDim();
 
-  if( val_marker != -1 )
+  /*--- If this zone has no parts of the marker, it will not participate
+   * in the first part of this function (collection). ---*/
+  if( val_marker != -1 ){
     nVertex  = geom->GetnVertex(  val_marker  );
-  else
+    nElems   = geom->GetnElem_Bound(val_marker);
+  } else {
     nVertex  = 0;
+    nElems   = 0;
+  }
 
+  /*--- Get the number of domain vertices on the marker, and a mapping
+  * (iVertex) -> (iLocalVertex, non-domain points being ignored). ---*/
+  unsigned long* iVertex_to_iLocalVertex = new unsigned long[ nVertex ];
+  nLocalVertex = 0;
+  for (iVertex = 0; iVertex < nVertex; iVertex++) {
+    iPoint = geom->vertex[val_marker][iVertex]->GetNode();
+    if (geom->nodes->GetDomain(iPoint)){
+      iVertex_to_iLocalVertex[iVertex] = nLocalVertex;
+      nLocalVertex++;
+    }  else {
+      iVertex_to_iLocalVertex[iVertex] = numeric_limits<unsigned long>::max();
+    }
+  }
 
-  su2double *Buffer_Send_Coord           = new su2double     [ nVertex * nDim ];
-  unsigned long *Buffer_Send_GlobalPoint = new unsigned long [ nVertex ];
-
-  unsigned long *Buffer_Send_nLinkedNodes       = new unsigned long [ nVertex ];
-  unsigned long *Buffer_Send_StartLinkedNodes   = new unsigned long [ nVertex ];
-  unsigned long **Aux_Send_Map                  = new unsigned long*[ nVertex ];
-
-#ifdef HAVE_MPI
-  int nProcessor = size, iRank;
-  unsigned long iTmp2, tmp_index, tmp_index_2;
-#endif
-
-  /*--- Copy coordinates and point to the auxiliar vector ---*/
+  // coordinates of all domain vertices on the marker
+  su2double *Buffer_Send_Coord           = new su2double     [ nLocalVertex * nDim ];
+  // global point IDs of all domain vertices on the marker
+  long *Buffer_Send_GlobalPoint = new long [ nVertex ];
 
-  nGlobalVertex     = 0;
-  nLocalVertex      = 0;
-  nLocalLinkedNodes = 0;
+  // Assign to each domain vertex on the marker, identified by local point ID,
+  // a set of surface-neighbor vertices on the marker, identified by global point ID.
+  map<unsigned long, forward_list<unsigned long>*> neighbors;
 
+  /*--- Define or initialize them. ---*/
   for (iVertex = 0; iVertex < nVertex; iVertex++) {
-
-    Buffer_Send_nLinkedNodes[iVertex] = 0;
-    Aux_Send_Map[iVertex]             = nullptr;
-
     iPoint = geom->vertex[val_marker][iVertex]->GetNode();
-
     if (geom->nodes->GetDomain(iPoint)) {
-      Buffer_Send_GlobalPoint[nLocalVertex] = geom->nodes->GetGlobalIndex(iPoint);
-
+      unsigned long iLocalVertex = iVertex_to_iLocalVertex[iVertex];
+      Buffer_Send_GlobalPoint[iLocalVertex] = (long) geom->nodes->GetGlobalIndex(iPoint);
```

Comment:
no C-style casts please, use `static_cast`

Reply:
ok, changed in 19d7a95

Reply:
in 612be75 I changed the type to `unsigned long` anyway